### PR TITLE
Simplify the ValuesAttribute

### PR DESCRIPTION
**Pull request by Bartleby2718**
This is my first time adding a new diagnostic, please take a closer look at Markdown file changes or strings that users can see. Thanks in advance!

### CODE REVIEW

File: src/nunit.analyzers/SimplifyValues/SimplifyValuesCodeFix.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+namespace NUnit.Analyzers.SimplifyValues;
+
+[ExportCodeFixProvider(LanguageNames.CSharp)]
+[Shared]
+public class SimplifyValuesCodeFix : CodeFixProvider
+{
+    internal const string SimplifyValuesTitle = "Simplify the Values attribute";
+
+    public override ImmutableArray<string> FixableDiagnosticIds
+        => ImmutableArray.Create(AnalyzerIdentifiers.SimplifyValues);
```
Comment (manfred-brands) [inline src/nunit.analyzers/SimplifyValues/SimplifyValuesCodeFix.cs:19]:
```suggestion
    public override ImmutableArray<string> FixableDiagnosticIds { get; }
        = ImmutableArray.Create(AnalyzerIdentifiers.SimplifyValues);
```

The original version would create a new instance every time it is called.
Yes, I know not all the Analyzer use this (yet)

**Review by manfred-brands — CHANGES_REQUESTED** [review]
Nice work, I only have some small comments.

File: src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzerConstants.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+namespace NUnit.Analyzers.SimplifyValues;
+
+internal static class SimplifyValuesAnalyzerConstants
+{
+    internal const string SimplifyValuesTitle = "Simplify the Values attribute";
+    internal const string SimplifyValuesMessage = "Simplify the Values attribute";
+    internal const string SimplifyValuesDescription = "Consider simplifying the ValuesAttribute when you're using the CombinatorialAttribute.";
```
[the remark this answers is not in the document]
Comment (Bartleby2718) [reply src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzerConstants.cs:7]: @manfred-brands I believe this should match https://github.com/nunit/nunit.analyzers/pull/758#discussion_r1634114854; which one do you prefer?

File: src/nunit.analyzers.tests/SimplifyValues/SimplifyValuesAnalyzerTests.cs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        RoslynAssert.Valid(this.analyzer, testCode);
+    }
+
+    [Test]
+    public void AnalyzeNotAllBooleanValuesAreInParameters()
+    {
+        var testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@"
+    public class AnalyzeNotAllBooleanValuesAreInParameters
+    {
+        public void Test([Values(new object[] { true })] bool b) { }
```
[the remark this answers is not in the document]
Comment (Bartleby2718) [reply src/nunit.analyzers.tests/SimplifyValues/SimplifyValuesAnalyzerTests.cs:137]: > Would anybody write such code?

I wouldn't, but I was just doing the same thing as in [`ValuesUsageAnalyzerTests.cs`](https://github.com/nunit/nunit.analyzers/blob/master/src/nunit.analyzers.tests/ValuesUsage/ValuesUsageAnalyzerTests.cs).

> add a test case for this

If I do, I get [CS0182: An attribute argument must be a constant expression, typeof expression or array creation expression of an attribute parameter type](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/parameter-argument-mismatch#arguments-for-attributes) for the test, so we can safely ignore this case.

### Would you like me to drop support for the array creation expression? I'd be happy to do that.

[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/SimplifyValues/SimplifyValuesAnalyzerTests.cs:137]: As you already did the work for the array creation expression, leave it in for now.

File: documentation/NUnit4001.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+| :--      | :--
+| Id       | NUnit4001
+| Severity | Info
+| Enabled  | True
+| Category | Style
+| Code     | [SimplifyValuesAnalyzer](https://github.com/nunit/nunit.analyzers/blob/master/src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzer.cs)
+
+## Description
+
+Consider simplifying the ValuesAttribute when you're using the CombinatorialAttribute.
```
[the remark this answers is not in the document]
Comment (Bartleby2718) [reply documentation/NUnit4001.md:15]: Closing in favor of https://github.com/nunit/nunit.analyzers/pull/758#discussion_r1634123798

File: src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        if (parameterSymbol is null)
+            return;
+
+        var methodSymbol = parameterSymbol.ContainingSymbol as IMethodSymbol;
+        if (methodSymbol is null)
+            return;
+
+        var methodAttributes = methodSymbol.GetAttributes();
+        var methodHasNonCombinatorialAttribute = methodAttributes
+            .Any(attribute => attribute.AttributeClass is { } attributeClass
```
Comment (manfred-brands) [inline src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzer.cs:68]: I find `is { }` very hard to read. 
```suggestion
            .Any(attribute => attribute.AttributeClass is INamedTypeSymbol attributeClass
```

File: src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                && StringComparer.OrdinalIgnoreCase.Equals(
```
Comment (manfred-brands) [inline src/nunit.analyzers/SimplifyValues/SimplifyValuesAnalyzer.cs:69]: We don't want 'nunit.framework' to match. 
```suggestion
                && StringComparer.Ordinal.Equals(
```

**Review by manfred-brands — APPROVED** [review]
Two nitpicks, nothing else.

**Comment by manfred-brands** [comment]
Thanks @Bartleby2718, I have merged the PR now.

**Comment by Bartleby2718** [comment]
@manfred-brands Thanks for providing feedback and merging the PR!

**Pull request merged**